packages/common/infra/src/di/core/collection.ts

```diff
@@ -0,0 +1,459 @@
+      cls as any,
+      dependenciesToFactory(cls, deps as any),
+      { scope: this.currentScope }
+    );
+
+    return this;
+  };
+
+  /**
+   * Add an implementation for identifier to the collection.
+   *
+   * @see {@link ServiceCollection}
+   *
+   * @example
+   * ```ts
+   * addImpl(ServiceIdentifier, ServiceClass, [dependencies, ...])
+   * or
+   * addImpl(ServiceIdentifier, Instance)
+   * or
+   * addImpl(ServiceIdentifier, Factory)
+   * ```
+   */
+  addImpl = <
+    Arg1 extends ServiceIdentifier<any>,
+    Arg2 extends Type<Trait> | ServiceFactory<Trait> | Trait,
+    Trait = ServiceIdentifierType<Arg1>,
+    Deps extends Arg2 extends Type<Trait>
+      ? TypesToDeps<ConstructorParameters<Arg2>>
+      : [] = Arg2 extends Type<Trait>
+      ? TypesToDeps<ConstructorParameters<Arg2>>
+      : [],
+    Arg3 extends Deps = Deps,
+  >(
+    identifier: Arg1,
+    arg2: Arg2,
+    ...[arg3]: Arg3 extends [] ? [] : [Arg3]
+  ): this => {
+    if (arg2 instanceof Function) {
+      this.collection.addFactory<any>(
+        identifier,
+        dependenciesToFactory(arg2, arg3 as any[]),
+        { scope: this.currentScope }
+      );
+    } else {
+      this.collection.addValue(identifier, arg2 as any, {
+        scope: this.currentScope,
+      });
+    }
+
+    return this;
+  };
+
+  /**
+   * same as {@link addImpl} but this method will override the service if it exists.
+   *
+   * @see {@link ServiceCollection}
+   *
+   * @example
+   * ```ts
+   * override(OriginServiceClass, NewServiceClass, [dependencies, ...])
+   * or
+   * override(ServiceIdentifier, ServiceClass, [dependencies, ...])
+   * or
+   * override(ServiceIdentifier, Instance)
+   * or
+   * override(ServiceIdentifier, Factory)
+   * ```
+   */
+  override = <
+    Arg1 extends ServiceIdentifier<any>,
+    Arg2 extends Type<Trait> | ServiceFactory<Trait> | Trait,
+    Trait = ServiceIdentifierType<Arg1>,
+    Deps extends Arg2 extends Type<Trait>
+      ? TypesToDeps<ConstructorParameters<Arg2>>
+      : [] = Arg2 extends Type<Trait>
+      ? TypesToDeps<ConstructorParameters<Arg2>>
+      : [],
+    Arg3 extends Deps = Deps,
+  >(
+    identifier: Arg1,
+    arg2: Arg2,
+    ...[arg3]: Arg3 extends [] ? [] : [Arg3]
+  ): this => {
+    if (arg2 instanceof Function) {
+      this.collection.addFactory<any>(
+        identifier,
+        dependenciesToFactory(arg2, arg3 as any[]),
+        { scope: this.currentScope, override: true }
+      );
+    } else {
+      this.collection.addValue(identifier, arg2 as any, {
+        scope: this.currentScope,
+        override: true,
+      });
+    }
+
+    return this;
+  };
+
+  /**
+   * Set the scope for the service registered subsequently
+   *
+   * @example
+   *
+   * ```ts
+   * const ScopeA = createScope('a');
+   *
+   * services.scope(ScopeA).add(XXXService, ...);
+   * ```
+   */
+  scope = (scope: ServiceScope): ServiceCollectionEditor => {
+    this.currentScope = scope;
```

Comment:
It looks like scope might be misused. E.g.:

```ts
// a.ts
import { services } from './services'
services.scope('foo').add(xxx)

// then in b.ts
import { services } from './services'
services.add(yyy); // may or may not be scope 'foo', depending on import order
```

If a service collection ref being used in two places, there might be an issue that the scope is changed but it is not known to the others.

Reply:
In fact in your example we always get

```
foo:
  xxx
root:
  yyy
```

when you call `add` / `addImpl` / `scope` / `override` on `ServiceCollection`, a `ServiceCollectionEditor` will be created, and your `scope("foo")` actually only acts on this `Editor` and not on `ServiceCollection`.

https://github.com/toeverything/AFFiNE/blob/fc2bcc8ad42c03c1ab1db9eb24550320773bdb28/packages/common/infra/src/di/core/collection.ts#L132-L155